src/tape_drivers/linux/sg/sg_tape.c

```diff
@@ -118,31 +121,25 @@ static inline int _parse_logPage(const unsigned char *logdata,
 								 const uint16_t param, uint32_t *param_size,
 								 unsigned char *buf, const size_t bufsize)
 {
-	uint16_t page_len, param_code, param_len;
-	uint32_t i;
+	const uint16_t page_len = ((uint16_t)logdata[2] << 8) | (uint8_t)logdata[3];
+	uint16_t param_code, param_len;
+	uint32_t i = LOG_PAGE_HEADER_SIZE;
 	uint32_t ret = -EDEV_INTERNAL_ERROR;
 
-	page_len = ((uint16_t)logdata[2] << 8) + (uint16_t)logdata[3];
-	i = LOG_PAGE_HEADER_SIZE;
-
 	while(i < page_len)
 	{
-		param_code = ((uint16_t)logdata[i] << 8) + (uint16_t)logdata[i+1];
+		param_code = ((uint16_t)logdata[i] << 8) | (uint8_t)logdata[i+1];
```

Comment:
I cannot understand why you change the cast from uint16_t to uint8_t. Do you have a problem on the precious code?

I believe `param_code` is `uint16_t`, so both variables, `logdata[i]` and `logdata[i+1]` shall be casted to `uint16_t` explicitly.

Reply:
Let's take a look to the original code.
```c
static inline int _parse_logPage(const unsigned char *logdata,
								 const uint16_t param, uint32_t *param_size,
								 unsigned char *buf, const size_t bufsize)
{
	uint16_t page_len, param_code, param_len;
	uint32_t i;
	uint32_t ret = -EDEV_INTERNAL_ERROR;

	page_len = ((uint16_t)logdata[2] << 8) + (uint16_t)logdata[3];
	i = LOG_PAGE_HEADER_SIZE;

	while(i < page_len)
	{
		param_code = ((uint16_t)logdata[i] << 8) + (uint16_t)logdata[i+1];
		param_len  = (uint16_t)logdata[i + LOG_PAGE_PARAMSIZE_OFFSET];
```

You are calling `logdata` as a `const unsigned char`, then all of the elements in that array are _8-bit_ (or _1 byte_) long. To me, the cast to `uint16_t` of the elements of `logdata[ ]` is totally *unnecessary*.

Let me explain my mind: If `page_len` is an `uint16_t` variable (_2 byte_ long), it can hold an `unsigned char` with a bitwise operation.

```c
page_len = (logdata[2] << 8) + logdata[3];
```

Let's assume the following:
```c
logdata[2] = 10001111;
logdata[3] = 00110011;
```

Then, the operation might look like this:
```c
page_len  = (0b10001111 << 8) | (0b00110011); 
                   // 10001111 00000

// then
page_len = 10001111 00110011;
```

With this being said, I only changed the cast to `uint8_t` in order to keep consistency with the code that you already developed. But in my opinion, the code should look like the following:

```c
static inline int _parse_logPage(const unsigned char *logdata,
								 const uint16_t param, uint32_t *param_size,
								 unsigned char *buf, const size_t bufsize)
{
	uint16_t page_len = (logdata[2] << 8) | logdata[3];
	uint16_t param_code, param_len;
	uint32_t i = LOG_PAGE_HEADER_SIZE;
	uint32_t ret = -EDEV_INTERNAL_ERROR;

	while(i < page_len)
	{
		param_code = (logdata[i] << 8) | logdata[i+1];
		param_len  = (uint16_t)logdata[i + LOG_PAGE_PARAMSIZE_OFFSET];
```

Reply:
No, you need to learn about C compiler and CPU more.

The point is the types `uintx_t` is not native types, but they are mapped to a correct native type by headers in `typedef` sentence. (see `stdint.h` on your system)

The code below is ambiguous because the computer cannot make a 'bit-wise OR' operation with different width variables, so `logdata[3]` is casted to `uint8_t (unsigned char)` and then implicitly casted to `unsigned short` (without `uint16_t`) in the `x86_64` architecture. Another behavior might be done in different architecture.

```C
((uint16_t)logdata[2] << 8) + (uint8_t)logdata[3];
```

In the previous code, this is clear description for C compiler.
```C
page_len = ((uint16_t)logdata[2] << 8) + (uint16_t)logdata[3];
```

Both `logdata[2]` and `logdata[3]` is casted to `uint16_t` first, make 'bit-wise OR' operation in 16-bit width and store the result to 16-bit wide page_len.